app/services/SubmissionService.scala

```diff
@@ -53,26 +53,45 @@ class SubmissionService @Inject()(
     }
   }
 
-  def submitPSub(nino: String, taxYears: Seq[TaxYearSelection], claimAmount: Int)
-               (implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Seq[HttpResponse]] = {
+  def submitPSub(nino: String, taxYears: Seq[TaxYearSelection], subscriptions: Map[Int, Seq[PSub]])
+                (implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Seq[HttpResponse]] = {
 
     getTaxYearsToUpdate(nino, taxYears).flatMap {
       claimYears =>
-        futureSequence(claimYears) {
-          taxYearSelection =>
-            val taxYear = TaxYearSelection.getTaxYear(taxYearSelection)
+
+        val psubsToUpdate: Seq[(Int, Seq[PSub])] = claimYears.flatMap {
+          year =>
+            subscriptions.get(getTaxYear(year)).filter(_.nonEmpty).map {
+              psubs =>
+                getTaxYear(year) -> psubs
+            }
+        }
+
+        futureSequence(psubsToUpdate) {
+          psubsByYear =>
+            val taxYear = psubsByYear._1
+            val claimAmount = claimAmountMinusDeductions(psubsByYear._2)
+
             taiService.updatePsubAmount(nino, taxYear, claimAmount)
         }
```

Comment:
or perhaps the whole thing can be reduced to something like :

```
claimYears =>
        Future.sequence(for {
          year <- claimYears
          psubs <- subscriptions.get(getTaxYear(year)).filter(_.nonEmpty)
        } yield taiService.updatePsubAmount(nino, getTaxYear(year), claimAmountMinusDeductions(psubs)))
```

Not sure if that's semantically the same.  Chat on Monday

Reply:
Wouldn't work due to needing to submit claims in a procedural manner due to etag incrementation.